ManagementPacks/Community.DataOnDemand/Scripts/Watch-TcpConnections.ps1

```diff
@@ -0,0 +1,539 @@
+            MIB_TCP_STATE_FIN_WAIT2,
+            MIB_TCP_STATE_CLOSE_WAIT,
+            MIB_TCP_STATE_CLOSING,
+            MIB_TCP_STATE_LAST_ACK,
+            MIB_TCP_STATE_TIME_WAIT,
+            MIB_TCP_STATE_DELETE_TCB
+        }
+        #endregion
+
+        #region APIs
+        // http://msdn2.microsoft.com/en-us/library/aa366073.aspx
+        private static class IPHelperAPI
+        {
+            [DllImport("iphlpapi.dll", SetLastError = true)]
+            internal static extern uint GetExtendedTcpTable(
+                IntPtr tcpTable,
+                ref int tcpTableLength,
+                bool sort,
+                int ipVersion,
+                TCP_TABLE_CLASS tcpTableType,
+                int reserved);
+        }
+        #endregion
+
+        #region wrapper class
+        private class IPHelperWrapper : IDisposable
+        {
+
+            public const int AF_INET = 2;    // IP_v4 = System.Net.Sockets.AddressFamily.InterNetwork
+            public const int AF_INET6 = 23;  // IP_v6 = System.Net.Sockets.AddressFamily.InterNetworkV6
+
+            // Creates a new wrapper for the local machine
+            public IPHelperWrapper() { }
+
+            // Disposes of this wrapper
+            public void Dispose() { GC.SuppressFinalize(this); }
+
+            public List<MIB_TCPROW_OWNER_PID> GetAllTCPv4Connections()
+            {
+                return GetTCPConnections<MIB_TCPROW_OWNER_PID, MIB_TCPTABLE_OWNER_PID>(AF_INET);
+            }
+
+            public List<MIB_TCP6ROW_OWNER_PID> GetAllTCPv6Connections()
+            {
+                return GetTCPConnections<MIB_TCP6ROW_OWNER_PID, MIB_TCP6TABLE_OWNER_PID>(AF_INET6);
+            }
+
+            public List<IPR> GetTCPConnections<IPR, IPT>(int ipVersion)
+            { //IPR = Row Type, IPT = Table Type
+
+                IPR[] tableRows;
+                int buffSize = 0;
+                FieldInfo dwNumEntriesField = typeof(IPT).GetField("dwNumEntries");
+
+                // how much memory do we need?
+                uint ret = IPHelperAPI.GetExtendedTcpTable(IntPtr.Zero, ref buffSize, true, ipVersion, TCP_TABLE_CLASS.TCP_TABLE_OWNER_PID_ALL, 0);
+                IntPtr tcpTablePtr = Marshal.AllocHGlobal(buffSize);
+
+                try
+                {
+                    ret = IPHelperAPI.GetExtendedTcpTable(tcpTablePtr, ref buffSize, true, ipVersion, TCP_TABLE_CLASS.TCP_TABLE_OWNER_PID_ALL, 0);
+                    if (ret != 0) return new List<IPR>();
+
+                    // get the number of entries in the table
+                    IPT table = (IPT)Marshal.PtrToStructure(tcpTablePtr, typeof(IPT));
+                    int rowStructSize = Marshal.SizeOf(typeof(IPR));
+                    uint numEntries = (uint)dwNumEntriesField.GetValue(table);
+
+                    // buffer we will be returning
+                    tableRows = new IPR[numEntries];
+
+                    IntPtr rowPtr = (IntPtr)((long)tcpTablePtr + 4);
```

Comment:
Hopefully the 4 here is a quantity of entries or something and not `sizeof(Int32)`, right? 😛